xmppserver/src/main/java/org/jivesoftware/admin/servlet/BlowfishMigrationServlet.java

```diff
@@ -0,0 +1,440 @@
+            throw new IllegalStateException("Encryption algorithm is " + encryptionAlgorithm +
+                    ", not Blowfish. Migration only applies to Blowfish.");
+        }
+
+        Log.info("Starting Blowfish migration from SHA1 to PBKDF2...");
+
+        // 2. Migrate XML properties FIRST (fail fast before database changes)
+        // This migrates encrypted properties in openfire.xml (e.g., database credentials)
+        int xmlMigrated = 0;
+        try {
+            xmlMigrated = JiveGlobals.migrateXMLPropertiesFromSHA1ToPBKDF2();
+            Log.info("Migrated {} XML properties from openfire.xml", xmlMigrated);
+        } catch (Exception e) {
+            // XML migration failed - stop before making any database changes
+            throw new RuntimeException("XML property migration failed. " +
+                    "No database changes were made. " + e.getMessage(), e);
+        }
+
+        // 3. Find all Blowfish-encrypted properties in database
+        List<EncryptedProperty> dbProperties = getAllEncryptedPropertiesFromDatabase();
+        Log.info("Found {} encrypted properties in database to migrate", dbProperties.size());
+
+        // 4. Get the master encryption key from JiveGlobals
+        String masterKey = JiveGlobals.getMasterEncryptionKey();
+        if (masterKey == null) {
+            Log.info("No custom encryption key configured - using default Blowfish key for migration");
+        }
+
+        // 5. Initialise both SHA1 and PBKDF2 encryptors
+        // We need two separate instances with different KDFs:
+        // - SHA1: for decrypting existing properties
+        // - PBKDF2: for re-encrypting with new KDF
+        Blowfish sha1Blowfish = new Blowfish();
+        sha1Blowfish.setKey(masterKey, JiveGlobals.BLOWFISH_KDF_SHA1);
+
+        Blowfish pbkdf2Blowfish = new Blowfish();
+        pbkdf2Blowfish.setKey(masterKey, JiveGlobals.BLOWFISH_KDF_PBKDF2);
+
+        // 6. Migrate each database property within a transaction
+        Connection con = null;
+        boolean originalAutoCommit = true;
+        int migrated = 0;
+        int failed = 0;
+        List<String> failedProperties = new ArrayList<>();
+
+        try {
+            con = DbConnectionManager.getConnection();
+            originalAutoCommit = con.getAutoCommit();
+            con.setAutoCommit(false); // Start transaction
+
+            for (EncryptedProperty prop : dbProperties) {
+                try {
+                    // Decrypt with SHA1-derived key
+                    String decrypted = sha1Blowfish.decryptString(prop.value);
+
+                    // Re-encrypt with PBKDF2-derived key
+                    String reencrypted = pbkdf2Blowfish.encryptString(decrypted);
+
+                    // Update property in database
+                    updateEncryptedProperty(con, prop.name, reencrypted);
+
+                    migrated++;
+
+                    if (migrated % 10 == 0) {
```

Comment:
This is a bit of a nit, but: The purpose of this code is to strike a balance between logging some progress and avoiding excessive log entries that can clutter log files. Logging based on an arbitrary count of elements often produces suboptimal results - sometimes the count is too high, other times too low.

What users generally want is a progress indicator when operations take longer than expected. With that in mind, it makes more sense to log periodically based on elapsed time rather than item count. This can be easily implemented, for example, with: 

```java
if (Duration.between(start, Instant.now()).compareTo(Duration.ofSeconds(5)) >= 0) {
    // log progress
}
```

This approach ensures that progress updates are consistent and meaningful, without overwhelming the logs.